pkg/api/raw.go

```diff
@@ -0,0 +1,108 @@
+	"io"
+	"net/http"
+
+	"github.com/gorilla/mux"
+
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/file/splitter"
+	"github.com/ethersphere/bee/pkg/jsonhttp"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+type rawPostResponse struct {
+	Hash swarm.Address `json:"hash"`
+}
+
+// rawUploadHandler handles upload of raw binary data of arbitrary length.
+func (s *server) rawUploadHandler(w http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	responseObject, err := s.splitUpload(ctx, r.Body, r.ContentLength)
+	if err != nil {
+		jsonhttp.BadGateway(w, err.Error())
+		return
+	}
+	jsonhttp.OK(w, responseObject)
+}
+
+func (s *server) splitUpload(ctx context.Context, r io.ReadCloser, l int64) (interface{}, error) {
+	chunkPipe := file.NewChunkPipe()
+	go func() {
+		data := make([]byte, swarm.ChunkSize)
+		c, err := io.CopyBuffer(chunkPipe, r, data)
+		if err != nil {
+			s.Logger.Debugf("raw: read count mismatch %d: %v", c, err)
+			s.Logger.Error("raw: read count mismatch")
+			return
+		}
+		if c != l {
+			s.Logger.Debugf("raw: read count mismatch %d: %v", c, err)
+			s.Logger.Error("raw: read count mismatch")
+			return
+		}
+		err = chunkPipe.Close()
+		if err != nil {
+			s.Logger.Errorf("raw: incomplete file write close %v", err)
+			s.Logger.Error("raw: incomplete file write close")
+		}
+	}()
+	sp := splitter.NewSimpleSplitter(s.Storer)
+	address, err := sp.Split(ctx, chunkPipe, l)
+	if err != nil {
+		s.Logger.Debugf("raw: split error %s: %v", address, err)
+		s.Logger.Error("raw: split error")
+		return errors.New("upload error"), nil
+	}
+	return rawPostResponse{Hash: address}, nil
+}
+
+// rawGetHandler handles retrieval of raw binary data of arbitrary length.
+func (s *server) rawGetHandler(w http.ResponseWriter, r *http.Request) {
+	addressHex := mux.Vars(r)["address"]
+	ctx := r.Context()
+
+	address, err := swarm.ParseHexAddress(addressHex)
+	if err != nil {
+		s.Logger.Debugf("raw: parse address %s: %v", addressHex, err)
+		s.Logger.Error("raw: parse address error")
+		jsonhttp.BadRequest(w, "invalid address")
+		return
+	}
+
+	j := joiner.NewSimpleJoiner(s.Storer)
+
+	dataSize, err := j.Size(ctx, address)
+	if err != nil {
+		if err == storage.ErrNotFound {
+			s.Logger.Debugf("raw: not found %s: %v", address, err)
+			s.Logger.Error("raw: not found")
+			jsonhttp.NotFound(w, "not found")
+			return
+		}
+		s.Logger.Debugf("raw: invalid root chunk %s: %v", address, err)
+		s.Logger.Error("raw: invalid root chunk")
+		jsonhttp.BadRequest(w, "invalid root chunk")
+		return
+	}
+
+	w.Header().Set("Content-Type", "application/octet-stream")
+	w.Header().Set("Content-Length", fmt.Sprintf("%d", dataSize))
+	err = file.JoinReadAll(j, address, w)
+	if err != nil {
+		s.Logger.Errorf("raw: data write %s: %v", address, err)
+		s.Logger.Error("raw: data input error")
+		jsonhttp.BadRequest(w, "failed to retrieve data")
```

Comment:
I haven't looked down the stack on this and how things are buffered. I guess the client will be simply faced with a "server premature connection close" if things go wrong. Probably ok, but as an end user I remember this as a bit of a pain. We should probably have solid logging here down the road.

Reply:
This will not close the connection. The issue is only with served content and go net/http refusing to write another header since one is already written.